scripts/build-pipelines.sh

```diff
@@ -0,0 +1,146 @@
+EOF
+
+printf "\n"
+
+# Using shards to optimise tests
+# In the future we can incorporate test durations rather than using
+# a static value for the parallel keyword
+
+cat << "EOF"
+build:linux:
+  stage: build
+  needs: []
+  parallel: 2
+  script:
+    - >
+        nix-shell --run '
+        npm run build --verbose;
+        npm test -- --ci --runInBand --coverage --shard=$CI_NODE_INDEX/$CI_NODE_TOTAL;
+        '
+  artifacts:
+    when: always
+    reports:
+      junit:
+        - ./tmp/junit/junit.xml
+      coverage_report:
+        coverage_format: cobertura
+        path: ./tmp/coverage/cobertura-coverage.xml
+    paths:
+      # Only the build:linux preserves the dist
+      - ./dist
+  coverage: '/All files[^|]*\|[^|]*\s+([\d\.]+)/'
+  rules:
+    # Runs on staging commits and ignores version commits
+    - if: $CI_COMMIT_BRANCH == 'staging' && $CI_COMMIT_TITLE !~ /^[0-9]+\.[0-9]+\.[0-9]+(?:-.*[0-9]+)?$/
+    # Runs on tag pipeline where the tag is a prerelease or release version
+    - if: $CI_COMMIT_TAG =~ /^v[0-9]+\.[0-9]+\.[0-9]+(?:-.*[0-9]+)?$/
+EOF
+
+printf "\n"
+
+cat << "EOF"
+build:windows:
+  stage: build
+  needs: []
+  parallel: 2
+  tags:
+    - windows
+  before_script:
+    - choco install nodejs --version=16.14.2 -y
+    - refreshenv
+  script:
+    - npm config set msvs_version 2019
+    - npm install --ignore-scripts
+    - $env:Path = "$(npm bin);" + $env:Path
+    - npm run build --verbose
+    - npm test -- --ci --runInBand --shard=$CI_NODE_INDEX/$CI_NODE_TOTAL
+  artifacts:
+    when: always
+    reports:
+      junit:
+        - ./tmp/junit/junit.xml
+  rules:
+    # Runs on staging commits and ignores version commits
+    - if: $CI_COMMIT_BRANCH == 'staging' && $CI_COMMIT_TITLE !~ /^[0-9]+\.[0-9]+\.[0-9]+(?:-.*[0-9]+)?$/
+    # Runs on tag pipeline where the tag is a prerelease or release version
+    - if: $CI_COMMIT_TAG =~ /^v[0-9]+\.[0-9]+\.[0-9]+(?:-.*[0-9]+)?$/
+EOF
+
+printf "\n"
+
+cat << "EOF"
+build:macos:
+  stage: build
+  needs: []
+  parallel: 2
+  tags:
+    - shared-macos-amd64
+  image: macos-11-xcode-12
+  variables:
+    HOMEBREW_NO_INSTALL_UPGRADE: "true"
+    HOMEBREW_NO_INSTALL_CLEANUP: "true"
+  before_script:
+    - eval "$(brew shellenv)"
+    - brew install node@16
+    - brew link --overwrite node@16
+    - hash -r
+  script:
+    - npm install --ignore-scripts
+    - export PATH="$(npm bin):$PATH"
+    - npm run build --verbose
+    - npm test -- --ci --runInBand --shard=$CI_NODE_INDEX/$CI_NODE_TOTAL
+  artifacts:
+    when: always
+    reports:
+      junit:
+        - ./tmp/junit/junit.xml
+  rules:
+    # Runs on staging commits and ignores version commits
+    - if: $CI_COMMIT_BRANCH == 'staging' && $CI_COMMIT_TITLE !~ /^[0-9]+\.[0-9]+\.[0-9]+(?:-.*[0-9]+)?$/
+    # Runs on tag pipeline where the tag is a prerelease or release version
+    - if: $CI_COMMIT_TAG =~ /^v[0-9]+\.[0-9]+\.[0-9]+(?:-.*[0-9]+)?$/
+EOF
```

Comment:
I'm not sure if these rules are useful. Remember the original trigger is the generating of the pipeline script. If it generated, then we expect to actually run it.

If it never generated it, then it'd never trigger the child pipeline, and therefore the child jobs don't need this either.

Same with all the other rules on the child pipeline.